utils/frame/try-runtime/cli/src/lib.rs

```diff
@@ -53,169 +79,310 @@ pub struct TryRuntimeCmd {
+			.runtime_code()?,
+		sp_core::testing::TaskExecutor::new(),
+	)
+	.execute(execution.into())
+	.map_err(|e| format!("failed to execute 'TryRuntime_on_runtime_upgrade' due to {:?}", e))?;
+
+	let (weight, total_weight) = <(u64, u64) as Decode>::decode(&mut &*encoded_result)
+		.map_err(|e| format!("failed to decode output due to {:?}", e))?;
+	log::info!(
+		"TryRuntime_on_runtime_upgrade executed without errors. Consumed weight = {}, total weight = {} ({})",
+		weight,
+		total_weight,
+		weight as f64 / total_weight as f64
+	);
+
+	Ok(())
 }
 
-fn parse_url(s: &str) -> Result<String, &'static str> {
-	if s.starts_with("ws://") || s.starts_with("wss://") {
-		// could use Url crate as well, but lets keep it simple for now.
-		Ok(s.to_string())
-	} else {
-		Err("not a valid WS(S) url: must start with 'ws://' or 'wss://'")
-	}
+async fn offchain_worker<B, ExecDispatch>(
+	shared: SharedParams,
+	command: OffchainWorkerCmd,
+	config: Configuration,
+)-> sc_cli::Result<()>
+where
+	B: BlockT,
+	B::Hash: FromStr,
+	B::Header: serde::de::DeserializeOwned,
+	<B::Hash as FromStr>::Err: Debug,
+	NumberFor<B>: FromStr,
+	<NumberFor<B> as FromStr>::Err: Debug,
+	ExecDispatch: NativeExecutionDispatch + 'static,
+{
+	let wasm_method = shared.wasm_method;
+	let execution = shared.execution;
+
+	let mut changes = Default::default();
+	let max_runtime_instances = config.max_runtime_instances;
+	let heap_pages = config.default_heap_pages;
+	let executor = NativeExecutor::<ExecDispatch>::new(
+		wasm_method.into(),
+		heap_pages,
+		max_runtime_instances,
+	);
+
+	let (mode, url) = match command.state {
+			State::Live {
+				url,
+				snapshot_path,
+				block_at,
+				modules
+			} => {
+				let online_config = OnlineConfig {
+					transport: url.to_owned().into(),
+					state_snapshot: snapshot_path.as_ref().map(SnapshotConfig::new),
+					modules: modules.to_owned().unwrap_or_default(),
+					at: block_at.as_ref()
+						.map(|b| b.parse().map_err(|e| format!("Could not parse hash: {:?}", e))).transpose()?,
+					..Default::default()
+				};
+
+				(Mode::Online(online_config), url)
+			},
+			State::Snap { snapshot_path } => {
+				// This is a temporary hack; the url is used just to get the header. We should try
+				// and get the header out of state, OR use an arbitrary header if thats ok, OR allow
+				// the user to feed in a header via file.
+				// This assumes you have a node running on local host default
+				let url = "ws://127.0.0.1:9944".to_string();
+				let mode = Mode::Offline(OfflineConfig {
+					state_snapshot: SnapshotConfig::new(snapshot_path),
+				});
+
+				(mode, url)
+			}
+	};
+
+	// get the code to inject inject into the ext
+	let spec = config.chain_spec;
+	let genesis_storage = spec.build_storage()?;
+	let code = StorageData(
+		genesis_storage
+			.top
+			.get(well_known_keys::CODE)
+			.expect("code key must exist in genesis storage; qed")
+			.to_vec(),
+	);
+	let code_key = StorageKey(well_known_keys::CODE.to_vec());
```

Comment:
Just noting if we don't want to override the runtime code we take out this block of code and the `inject` on the builder